pymongoexplain/commands.py

```diff
@@ -110,18 +122,30 @@ def __init__(self, collection: Collection,
         super().__init__(collection.name)
         for key, value in kwargs.items():
             self.command_document[key] = value
+        if self.command_document["filter"] == {}:
+            self.command_document = {}
         self.command_document = convert_to_camelcase(self.command_document)
 
     @property
     def command_name(self):
         return "find"
 
+
 class FindAndModifyCommand(BaseCommand):
     def __init__(self, collection: Collection,
                  kwargs):
         super().__init__(collection.name)
+        print(kwargs)
         for key, value in kwargs.items():
-            self.command_document[key] = value
+            if key == "hint":
+                self.command_document["hint"] = value if \
+                    isinstance(value, str) else SON(value)
+            else:
+                self.command_document[key] = value
+        if "replacement" in self.command_document.keys():
+            self.command_document["update"] = self.command_document[
+                "replacement"]
+            del self.command_document["replacement"]
```

Comment:
I think it would be more readable to construct the command correctly from the start rather than constructing an invalid command and then patching it up after.

Reply:
Confused about this comment. The command is being constructed in this step. The kwargs value of replacement has to be patched in that way otherwise it will simply not produce the correct output. If we did not put this code there, then it would have to be duplicated multiple times in explainable_collection.py

Reply:
There are certain processing steps that must be followed to match pymongo's output, and most of them seem rather arbitrary so the only way for me to verify is to match my code to the test cases.

Reply:
What I'm trying to get at is that it's confusing to add the "replacement" field to command_document and then immediately remove it. Instead I think this class should handle all the findAndModify specific options while we constructing the command. Like this:
```
    command_document = {
        "findAndModify": collection_name,
        "query": filter,
        "new":  return_document,
        "update": replacement,
    }
    collation = validate_collation_or_none(kwargs.pop('collation', None))
    if collation is not None:
        cmd["collation"] = collation
    if projection is not None:
        cmd["fields"] = helpers._fields_list_to_dict(projection,
                                                     "projection")
    if sort is not None:
        cmd["sort"] = helpers._index_document(sort)
    if upsert is not None:
        common.validate_boolean("upsert", upsert)
        cmd["upsert"] = upsert
    if hint is not None:
        if not isinstance(hint, str):
            hint = helpers._index_document(hint)
        cmd['hint'] = hint
    if array_filters is not None:
        cmd["arrayFilters"] = array_filters
```

This approach exposes numerous bugs in the current approach. Sometimes an argument name needs validation (like a type assertion), sometimes it needs to be transformed (like `collation` or `hint`), and sometimes the python argument name is different from the command argument name (like `replacement` -> `update`).

Reply:
I removed the deleting statements and tried to condense all the logic down into the for loops.